### PR TITLE
Ruida UDP comms hardening.

### PR DESCRIPTION
**Pull request by StevenIsaacs**
This PR represents the next step in comms hardening for UDP communications with a Ruida controller. It replaces the receive thread with a state driven send and receive (handshake) thread to more gracefully handle failure conditions and reply data from the controller. And, because it is necessary for the handshake to know whether or not a reply is expected, the swizzle has been moved to the handshake thread.

Sending data to the controller is now double buffered by virtue of a queue between the write method and the handshake thread. This queue also serves to provide sync and locks until space is available. The queue size is large enough to avoid locking in most cases so the mechanism is mostly precautionary.

The new handshake thread has been tested and performs as expected. There may be a corner case or two which may reveal a problem but otherwise it is solid. Simple cut files are working with no problems. On the other hand, complex files such as a material tests are now triggering a new manifestation of the "invalid file" problem which has been determined to be unrelated to comms.

NOTE: Current testing has been with the laser disabled so power setting and speed parameters are not yet known to be correct.

NOTE: Some of these changes may be applicable to other interfaces (e.g. USB/serial) but have not been considered at this time.

### CODE REVIEW

File: meerk40t/ruida/udp_connection.py [hunk 1]
```diff
@@ -54,20 +72,151 @@ def abort_connect(self):
...
+        while _tries:
+            try:
+                self.send_q.put(data, timeout=self._to)
+                break
+            except queue.Full:
+                _tries -= 1
+                if _tries == 0:
+                    self.service.signal("warning", "Ruida", "Send queue FULL")
+                continue
+        self.send(data) # TODO: Where this goes is not known at this time.
```
[the remark this answers is not in the document]
Comment (StevenIsaacs) [reply meerk40t/ruida/udp_connection.py:95]: This send is not related to comms as far as I can tell. It appears to send to a channel. I haven't stepped this with the debugger to know for sure. Thus, the comment.

File: meerk40t/ruida/udp_connection.py [hunk 2]
```diff
@@ -54,20 +72,151 @@ def abort_connect(self):
...
+                        # TODO: Need to understand how the address can be used
+                        # further connection sanity checking.
+                        self.recv_address = _address
+                    _ack = self.unswizzle(_data)
+                    if len(_ack) == 1:
+                        if _ack == ACK:
+                            # Signal that the next message can be sent.
+                            _ack_pending = False
+                            self.acks += 1
+                        # TODO: Add resend of _message if not ACK.
```
[the remark this answers is not in the document]
Comment (StevenIsaacs) [reply meerk40t/ruida/udp_connection.py:191]: Yes, I know and it's why I placed the TODO there. The AI suggested code won't work.

File: meerk40t/ruida/udp_connection.py [hunk 3]
```diff
@@ -54,20 +72,151 @@ def abort_connect(self):
...
+                    else:
+                        self.events('Reply data when expecting ACK.')
+                        # Reply data in response to a command. Forward to be
+                        # processed.
+                        self.replies += 1
+                        _reply_pending = False
+                        # TODO: May need a way to restore sync.
+                        self.recv(_reply) # Just in case
```
[the remark this answers is not in the document]
Comment (StevenIsaacs) [reply meerk40t/ruida/udp_connection.py:199]: This is indeed a possibility. That should have been _ack. I'll add it to the list for the next round.

File: meerk40t/ruida/udp_connection.py [hunk 4]
```diff
@@ -54,20 +72,151 @@ def abort_connect(self):
...
+                        _data, _address = self.socket.recvfrom(1024)
+                        self.recv_address = _address
+                        # TODO: Add address sanity check?
+                        _reply_pending = False
+                        self.replies += 1
+                        _reply = self.unswizzle(_data)
+                        self.recv(_reply) # Forward to client.
+                    except socket.error as e:
+                        if (e.errno == socket.EWOULDBLOCK or
+                            e.errno == socket.EAGAIN):
```
[the remark this answers is not in the document]
Comment (StevenIsaacs) [reply meerk40t/ruida/udp_connection.py]: hmmm... two AIs disagree on this one. Gemini suggested the first but I'll go with the second.
